Implements the image-cropper package in the editor.

ramonjd opened the pull request:
Relies on https://github.com/WordPress/gutenberg/pull/72414

## What? How?

Implements the [image-cropper package](https://github.com/WordPress/gutenberg/pull/72414) in the editor.

https://github.com/user-attachments/assets/9c7e854b-a618-4aff-b174-0f72ec6a4e7d

## Why?

The image cropper component and hooks have been moved to their own package so they can be used across the app, and not just in side the block editor.

## Testing

Insert an image block in the editor and crop the image. Make sure to rotate, zoom, and select different aspect ratios.

Save different crops.

There should be no regressions in the way images are cropped in the editor.

There are 2 main differences:

- When rotating, if the user has chosen a non-default aspect ratio from the drop down, the aspect ratio will persist 
- When rotating, if the user has zoomed, the aspect ratio will persist

ramonjd commented on packages/block-editor/src/components/image-editor/index.js:
```diff
@@ -26,36 +27,40 @@ export default function ImageEditor( {
-		>
-			<Cropper
-				borderProps={ borderProps }
+		<ImageCropperProvider>
```
The only difference is wrapping the image cropper provider here.

tellthemachines replied:
So will we always need the two providers or will it be possible to consolidate at some point?

ramonjd replied:
Good question. Consolidation is possible. I started off with only one provider, but it's a hefty refactor since the context contains save/transforms, and some bespoke rotation callback.

For that reason i thought it best just to start using the library first, then do the refactor later.

ramonjd commented on packages/block-editor/src/components/image-editor/cropper.js:
```diff
@@ -60,24 +48,7 @@ export default function ImageCropper( {
-					setZoom( newZoom * 100 );
-				} }
-			/>
+			<ImageCropperComponent src={ editedUrl || url } />
```
The image cropper component handles all the attributes internally

ramonjd commented on packages/block-editor/src/components/image-editor/use-transform-image.js:
```diff
@@ -1,39 +1,73 @@
+	 * This creates a new rotated image file, bypassing the image-cropper’s CSS transform rotation.
+	 * It's a bespoke solution to ensure that the rotated image fills the entire canvas.
+	 */
+	const [ iternalRotation, setInternalRotation ] = useState( 0 );
```
All this is to prevent regression. The current implementation bypasses the cropper package's rotation altogether by rotating a canvas blob image and reloading the cropper. 

This is presumably so the image can fill the canvas area.

tellthemachines commented on packages/image-cropper/src/components/image-cropper/index.tsx:
Not sure why we're still seeing this diff? The image-cropper package should be in trunk already.

andrewserong replied:
Hrm, might need a fresh rebase?

tellthemachines reviewed (commented):
The code changes specific to this PR LGTM but not sure why the image-cropper package is still appearing in the diff.

One thing I noticed while testing, which is unrelated to this PR as it also happens on trunk, is that when applying the cropper changes a notification pops up with an "undo" button. Clicking on that undo is irreversible, because it doesn't seem to be in sync with the post "undo" and "redo" buttons. Would be worth fixing separately.

ramonjd committed:
Add image cropper integration to block editor

- Added `@wordpress/image-cropper` dependency to package.json and tsconfig.json.
- Replaced `react-easy-crop` with the new `ImageCropperComponent` in cropper.js.
- Wrapped ImageEditor component with `ImageCropperProvider` for context management.
- Updated useTransformImage hook to utilize the new image cropper state management.
- Adjusted ZoomDropdown to handle zoom values correctly.

This update enhances the image editing capabilities within the block editor by integrating a more robust image cropping solution.

ramonjd committed:
Update package-lock.json to include new image-cropper package

tellthemachines commented on packages/block-editor/src/components/image-editor/use-transform-image.js:
```diff
@@ -88,33 +127,40 @@ export default function useTransformImage( {
 		}
-	}, [ rotation, defaultAspect, url ] );
+	}, [
+		iternalRotation,
```
Should this be `internalRotation`?

ramonjd replied:
The `eternalRotation`

ramonjd committed:
All hail the eternal rotation!

tellthemachines commented on packages/block-editor/src/components/image-editor/use-transform-image.js:
```diff
@@ -1,39 +1,73 @@
+	 * rotateClockwise rotates the image by 90° clockwise by drawing the original image onto a canvas with rotation applied,
+	 * then saves it as a new blob URL (editedUrl).
+	 * This creates a new rotated image file, bypassing the image-cropper’s CSS transform rotation.
+	 * It's a bespoke solution to ensure that the rotated image fills the entire canvas.
```
It seems to fill the content width rather than the entire canvas. I'm not sure it should do either of those things, but this is consistent with behaviour in trunk for now so fine for this PR.

ramonjd replied:
Fair point, best to make it clearer. I updated the comment. 

It was existing behaviour that took me a bit to grok so I was hoping to leave a helpful comment for the next folks! 😄 

Thanks!

tellthemachines reviewed (approved):
LGTM! Preserving the set aspect ratio on rotation is a nice improvement.

ramonjd committed:
Update comment to clarify that the rotated image fills the content width.

andrewserong reviewed (commented):
One tiny detail: now that the image cropper package is being imported into the block-editor package, I think this means we'll need to remove the `"private": true` flag in the image cropper package (based on the discussion in https://github.com/WordPress/gutenberg/pull/73633#discussion_r2570477484):

https://github.com/WordPress/gutenberg/blob/038f2f16d91161cc33925cba478b9476d9bf101a/packages/image-cropper/package.json#L48

andrewserong reviewed (approved):
Code change otherwise looks good, and testing well for me, too. Nice work! 🚀

ramonjd committed:
Remove private flag from image-cropper package.json and ensure public access for publishing.

ramonjd committed:
Add @wordpress/image-cropper entry to manifest.json

- Included new package entry for @wordpress/image-cropper in the manifest.
- Updated markdown source path to point to the package's README.md.

This addition enhances the documentation and visibility of the image cropper package within the project.

ramonjd committed:
Merge remote-tracking branch 'origin/update/integrate-image-cropper-package' into update/integrate-image-cropper-package
